### PR TITLE
Run type checker against all languages

### LINKED ISSUE #7801: Compile output from SDK generator tests
`TestGeneratePackage` for each of the language SDK generators and the common test infrastructure in [`TestSDKCodegen`](https://github.com/pulumi/pulumi/blob/2b97340133682a3e055c4fce6b4adcbd0783cb99/pkg/codegen/internal/test/sdk_driver.go#L106) only does string matching as verification. It does not exercise the generated code or even make sure that the SDK compiles. 

Unfortuantely, it seems like the default way that we write these tests makes to overlook errors as we generate test output:

1. Author test schema
2. Write test harness
3. Run test suite with `PULUMI_ACCEPT=true go test ./...` to generate files to disk.

No where in this flow is process to make sure that the generated code works or compiles. 

We should add an additional step at the end of these tests to use language specific tooling, to compile, typecheck, or build whatever artifacts are appropriate for each supported language.

Follow up action item from errors introduced in https://github.com/pulumi/pulumi/pull/7059 related to https://github.com/pulumi/pulumi/issues/7595

### PR DESCRIPTION
**Pull request by iwahbe**
# Description

Fixes #7801

I add tests for each generated SDK. 
- [X] `nodejs`
  This is implemented by invoking the `ts-node` cli over `npm`. I think this is the right approach, as we only have one program looking over our `.ts` code.
- [X] `python`
  We don't type check, only parse. This is done with the builtin module `py_compile`.
- [X] `dotnet`
  Currently, we invoke `dotnet build`. If you has a better (faster) solution, please let me know. 
- [X] `go`
  We also build go. This bother's me less, as its fast. 
- [ ] docs
I'm not sure what tests to run. If you can think of any, let me know. 

## Checklist

- [X] I have added tests that prove my fix is effective or that my feature works
  That's more or less the feature. 

- [ ] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

**Comment by iwahbe** [comment]
Given that this is purely a change in testing, I don't think that this needs a changelog entry.

File: pkg/codegen/internal/test/testdata/provider-config-schema/schema.json [hunk 1]
```diff
@@ -23,10 +23,7 @@
         "type": "string",
         "description": "This is a huge secret",
         "defaultInfo": {
```
Comment (t0yv0) [inline pkg/codegen/internal/test/testdata/provider-config-schema/schema.json:25]: Any special reason for these formatting changes? Removing them would make review easier :)

File: pkg/codegen/internal/test/testdata/provider-config-schema/dotnet/Pulumi.Configstation.csproj [hunk 2]
```diff
@@ -40,6 +40,7 @@
   </ItemGroup>
 
   <ItemGroup>
+    <PackageReference Include="Pulumi" Version="3.12" />
```
Comment (t0yv0) [inline pkg/codegen/internal/test/testdata/provider-config-schema/dotnet/Pulumi.Configstation.csproj:43]: Now ideally we'd compile against the version in the repo, not published version 3.12. Is this the case?

Comment (t0yv0) [reply pkg/codegen/internal/test/testdata/provider-config-schema/dotnet/Pulumi.Configstation.csproj:43]: That'd be ProjectReference.. but perhaps this is substituted dynamically?

File: pkg/codegen/internal/test/sdk_driver.go [hunk 3]
```diff
@@ -30,16 +38,19 @@ var sdkTests = []sdkTest{
 		Description: "Simple schema with a two part name (foo-bar)",
 	},
 	{
-		Directory:   "external-resource-schema",
-		Description: "External resource schema",
+		Directory:        "external-resource-schema",
+		Description:      "External resource schema",
+		SkipCompileCheck: codegen.NewStringSet(nodejs, golang),
```
Comment (t0yv0) [inline pkg/codegen/internal/test/sdk_driver.go:43]: These are broken? :)

File: pkg/codegen/python/gen_test.go [hunk 4]
```diff
@@ -57,7 +59,26 @@ func TestRelPathToRelImport(t *testing.T) {
 }
 
 func TestGeneratePackage(t *testing.T) {
-	test.TestSDKCodegen(t, "python", GeneratePackage)
+	test.TestSDKCodegen(t, "python", GeneratePackage, typeCheckGeneratedPackage)
+}
+
+// We can't type check a python program. We just check for syntax errors.
```
Comment (t0yv0) [inline pkg/codegen/python/gen_test.go:65]: We kind of can with mypy but it's not been working very well. I think this is a fine choice here 👍

**Comment by t0yv0** [comment]
Awesome. I assume there's a lot of busy work here trying to make the examples compile that weren't really designed to compile. Please @iwahbe feel free to skip more of these or spread the work around..

File: pkg/codegen/internal/test/testdata/provider-config-schema/dotnet/Pulumi.Configstation.csproj (hunk 2 above)
Comment (iwahbe) [reply pkg/codegen/internal/test/testdata/provider-config-schema/dotnet/Pulumi.Configstation.csproj:43]: These files are generated from a Pulumi schema: `schema.json`. I don't think that we support `ProjectReference` (check out `codegen/dotnet/gen.go:1986`).

Comment (iwahbe) [reply pkg/codegen/internal/test/testdata/provider-config-schema/dotnet/Pulumi.Configstation.csproj:43]: https://github.com/pulumi/pulumi/issues/7957

File: pkg/codegen/internal/test/testdata/provider-config-schema/schema.json (hunk 1 above)
Comment (iwahbe) [reply pkg/codegen/internal/test/testdata/provider-config-schema/schema.json:25]: Sorry. That was a "format on save" behavior. It won't happen again. I can undo it if desired. 

As an aside: how do you do code reviews that involve programatic changes to 100+ files.

File: pkg/codegen/internal/test/sdk_driver.go (hunk 3 above)
Comment (iwahbe) [reply pkg/codegen/internal/test/sdk_driver.go:43]: Unfortunately. 
```
X go get
# github.com/pulumi/pulumi/pkg/codegen/internal/test/testdata/external-resource-schema/typecheck/go/example
./component.go:85:24: undefined: "github.com/pulumi/pulumi-kubernetes/sdk/v3/go/kubernetes/meta/v1".ObjectMetaArrayInput
./component.go:86:24: undefined: "github.com/pulumi/pulumi-kubernetes/sdk/v3/go/kubernetes/meta/v1".ObjectMetaMapInput
./component.go:88:24: undefined: "github.com/pulumi/pulumi-kubernetes/sdk/v3/go/kubernetes/meta/v1".ObjectMetaArrayInput
./component.go:89:24: undefined: "github.com/pulumi/pulumi-kubernetes/sdk/v3/go/kubernetes/meta/v1".ObjectMetaMapInput
```

```
⨯ Unable to compile TypeScript:
component.ts(86,66): error TS2694: Namespace '"/Users/ianwahbe/go/src/github.com/pulumi/pulumi/pkg/codegen/internal/test/testdata/external-resource-schema/typecheck/nodejs/node_modules/@pulumi/kubernetes/types/input".meta.v1' has no exported member 'ObjectMetaArgs'.
component.ts(87,84): error TS2694: Namespace '"/Users/ianwahbe/go/src/github.com/pulumi/pulumi/pkg/codegen/internal/test/testdata/external-resource-schema/typecheck/nodejs/node_modules/@pulumi/kubernetes/types/input".meta.v1' has no exported member 'ObjectMetaArgs'.
component.ts(88,98): error TS2694: Namespace '"/Users/ianwahbe/go/src/github.com/pulumi/pulumi/pkg/codegen/internal/test/testdata/external-resource-schema/typecheck/nodejs/node_modules/@pulumi/kubernetes/types/input".meta.v1' has no exported member 'ObjectMetaArgs'.
component.ts(89,73): error TS2694: Namespace '"/Users/ianwahbe/go/src/github.com/pulumi/pulumi/pkg/codegen/internal/test/testdata/external-resource-schema/typecheck/nodejs/node_modules/@pulumi/kubernetes/types/input".meta.v1' has no exported member 'ObjectMetaArgs'.
component.ts(90,91): error TS2694: Namespace '"/Users/ianwahbe/go/src/github.com/pulumi/pulumi/pkg/codegen/internal/test/testdata/external-resource-schema/typecheck/nodejs/node_modules/@pulumi/kubernetes/types/input".meta.v1' has no exported member 'ObjectMetaArgs'.
component.ts(91,105): error TS2694: Namespace '"/Users/ianwahbe/go/src/github.com/pulumi/pulumi/pkg/codegen/internal/test/testdata/external-resource-schema/typecheck/nodejs/node_modules/@pulumi/kubernetes/types/input".meta.v1' has no exported member 'ObjectMetaArgs'.
```

File: pkg/codegen/go/gen_test.go [hunk 5]
```diff
@@ -65,7 +68,51 @@ func TestGeneratePackage(t *testing.T) {
...
+	root := pwd
+	if len(dir) == 1 {
+		require.True(t, dir[0].IsDir())
+		root = filepath.Join(pwd, dir[0].Name())
+	}
+	t.Logf("*** Testing go in dir: %q ***", root)
+
+	var stdout, stderr bytes.Buffer
+	cmdOptions := integration.ProgramTestOptions{Stderr: &stderr, Stdout: &stdout}
+	// We don't want to corrupt the global go.mod and go.sum with packages we
```
Comment (t0yv0) [inline pkg/codegen/go/gen_test.go:96]: Huh interesting, do I understand correctly that codegen does not emit a go.mod file? So we need to do it here? SGTM in principle.

File: pkg/codegen/internal/test/testdata/external-resource-schema/go/example/argFunction.go [hunk 6]
```diff
@@ -7,7 +7,7 @@ import (
 	"context"
 	"reflect"
 
-	"github.com/pulumi/pulumi-random/sdk/v2/go/random"
+	"github.com/pulumi/pulumi-random/sdk/v4/go/random"
```
Comment (t0yv0) [inline pkg/codegen/internal/test/testdata/external-resource-schema/go/example/argFunction.go:10]: There's some nastiness here.. usually when I get these diffs I re-run with PULUMI_IGNORE_AMBIENT_PLUGINS=1 and I don't get those anymore; however your PR seems to have passed the CI checks so maybe these are the right diffs that have to be included.

File: pkg/codegen/internal/test/testdata/provider-config-schema/schema.json (hunk 1 above)
Comment (t0yv0) [reply pkg/codegen/internal/test/testdata/provider-config-schema/schema.json:25]: NP at all.. It's a tiny thing. If we agree on the spec of the right "format on save" to use that'd be nice, reformat and avoid the noise.

File: pkg/codegen/nodejs/gen_test.go [hunk 7]
```diff
@@ -2,15 +2,61 @@
...
+		Verbose: true,
+		Stderr:  &stderr,
+		Stdout:  &stdout,
+	}
+
+	// TODO remove when https://github.com/pulumi/pulumi/pull/7938 lands
+	file := filepath.Join(pwd, "package.json")
+	oldFile, err := ioutil.ReadFile(file)
+	require.NoError(t, err)
+	newFile := strings.ReplaceAll(string(oldFile), "${VERSION}", "0.0.1")
```
Comment (t0yv0) [inline pkg/codegen/nodejs/gen_test.go:40]: Interesting I'm doing smth similar in my WIP PR and didn't seem to need it, I'll double check.

File: pkg/codegen/nodejs/gen_test.go [hunk 8]
```diff
@@ -2,15 +2,61 @@
...
+	err = ioutil.WriteFile(file, []byte(newFile), 0600)
+	require.NoError(t, err)
+
+	err = integration.RunCommand(t, "npm install", []string{npm, "i"}, pwd, &cmdOptions)
+	require.NoError(t, err)
+	err = integration.RunCommand(t, "typecheck ts",
+		[]string{filepath.Join(".", "node_modules", ".bin", "tsc"), "--noEmit"}, pwd, &cmdOptions)
```
Comment (t0yv0) [inline pkg/codegen/nodejs/gen_test.go:47]: Nice!

File: pkg/codegen/nodejs/gen_test.go [hunk 9]
```diff
@@ -2,15 +2,61 @@
...
 
 func TestGeneratePackage(t *testing.T) {
-	test.TestSDKCodegen(t, "nodejs", GeneratePackage)
+	test.TestSDKCodegen(t, "nodejs", GeneratePackage, typeCheckGeneratedPackage)
+}
+
+func typeCheckGeneratedPackage(t *testing.T, pwd string) {
+	var err error
+	var npm string
+	npm, err = executable.FindExecutable("npm")
```
Comment (t0yv0) [inline pkg/codegen/nodejs/gen_test.go:26]: There's also yarn. I think the rest of the codebase uses `yarn link @pulumi/pulumi` to test against in-repo Node SDK.

**Review by t0yv0 — APPROVED** [review]
l love this. I think this should go in so we start getting more confidence out of our  I've eyeballed the changes and LGTM. The comments I have around building against in-source SDK and using yarn these can go in follow up PR, in fact let me open my WIP PR where I am wrestling the Node part of this equation that has some of these ideas.

**Comment by t0yv0** [comment]
Linking in my WIP https://github.com/pulumi/pulumi/pull/7971

File: pkg/codegen/go/gen_test.go (hunk 5 above)
Comment (iwahbe) [reply pkg/codegen/go/gen_test.go:96]: Codegen does not emit a `go.mod` file. We have a very strange build tree for go projects by default. I think these are related, but it's not clear. 
```
pulumi/sdk
├── dotnet
├── go
├── go.mod
├── go.sum
├── nodejs
├── proto
├── python
└── README.md
```

File: pkg/codegen/nodejs/gen_test.go (hunk 7 above)
Comment (iwahbe) [reply pkg/codegen/nodejs/gen_test.go:40]: `npm` needs this but `yarn` does not. I figured that I should test on the most finicky version.

**Pull request merged**